Add Ruby/Basics exercise (#1043)
exercism/v3: languages/ruby/exercises/concept/basics/.docs/introduction.md

> Ruby is a dynamic [object-oriented language]. Everything in Ruby is an [object].

Reviewer: ```suggestion
Ruby is a dynamic [object-oriented language][object-oriented-programming]. Everything in Ruby is an [object][object].
```

There is no link defined for the `[object]` reference, so one has to be added to the bottom of this file.

> There are two ways to assign objects to names in Ruby - assigning variables or constants. Variables always start with lower-case letters and use "snake case" for their formatting. A variable can be redefined to different objects over its lifetime. For example, `my_first_variable` can be defined and redefined many times using the `=` operator:

Reviewer: Maybe define the snake casing in the description itself and link to the [Snake case Wikipedia page](https://en.wikipedia.org/wiki/Snake_case)?

```suggestion
There are two ways to assign objects to names in Ruby - assigning variables or constants. Variables always start with lower-case letters and use [snake_case][snake_case] for their formatting. A variable can be redefined to different objects over its lifetime. For example, `my_first_variable` can be defined and redefined many times using the `=` operator:
```

```ruby
my_first_variable = 1
my_first_variable = "Some string"
my_first_variable = SomeComplexObject.new
```

> Constants, however, can only be assigned once. The must start with captial letters and are normally written in block capitals with words seperated by underscores. For example:

Reviewer: ```suggestion
Constants, however, can only be assigned once. The must start with capital letters and are normally written in block capitals with words separated by underscores. For example:
```

> Constants, however, can only be assigned once. The must start with capital letters and are normally written in block capitals with words separated by underscores. For example:

Reviewer: ```suggestion
Constants, however, can only be assigned once. They must start with capital letters and are normally written in block capitals with words separated by underscores. For example:
```

Ruby is a dynamic [object-oriented language]. Everything in Ruby is an [object][object].

> There are two ways to assign objects to names in Ruby - assigning variables or constants. Variables always start with lower-case letters and use [snake case][snake-case] for their formatting. A variable can be redefined to different objects over its lifetime. For example, `my_first_variable` can be defined and redefined many times using the `=` operator:

Reviewer 2: ```suggestion
There are a few ways to assign objects to names in Ruby - assigning  local variables, (we will refer to these generally as _variables_ or in the singular, _variable_) or constants. Variables always start with lower-case letters and use [snake case][snake-case] for their formatting. A variable can reference different objects over its lifetime. For example, `my_first_variable` can be defined and redefined many times using the `=` operator:
```

Author: @kotp I'm not sure I understand these changes. Help me understand pls :)
1) If we change "two" to "a few" we should provide more than two examples. I'm not sure what other ways are though?
2) Why the reference to `_local_` and the content in brackets? Why can't we just say "variables"?

Reviewer 2: 1. There are a few, we will show two.  (because the other ways don't really matter at this point)
2. Because there are instance variables, class variables, global variables.  The italics are for emphasis that we are talking about specific variables here.  They start with a lower case letter, others start with a sigil.  We don't have to mention it, but we can suggest that we are talking about a specific thing by emphasizing the term.

Author: Ok - I think this works:

```suggestion
There are two primary ways to assign objects to names in Ruby - using variables or constants. Variables are always written in [snake case][snake-case]. A variable can reference different objects over its lifetime. For example, `my_first_variable` can be defined and redefined many times using the `=` operator:
```

Reviewer 2: Yes, I agree... keep the white lies to a minimum, but when necessary, keep them invisible as well.

```ruby
my_first_variable = 1
my_first_variable = "Some string"
my_first_variable = SomeComplexObject.new
```

> Constants, however, can only be assigned once. They must start with capital letters and are normally written in block capitals with words separated by underscores. For example:

Reviewer 2: ```suggestion
Constants, however, are meant to be assigned once. They must start with capital letters and are normally written in block capitals with words separated by underscores. For example:
```

```ruby
MY_FIRST_CONSTANT = 10

# Redefining not allowed
# MY_FIRST_CONSTANT = "Some String"
```

> Ruby is organised into classes. Classes are defined using the `class` keyword followed by the name of the class. Classes are initialized using the `.new` method.

Reviewer 2: ```suggestion
Ruby is organised into classes. Classes are defined using the `class` keyword followed by the name of the class. Classes are initialized, or _instantiated_, using the `new` method.
```

Reviewer 3: Why do we need underscores for the `instantiated`?

Author: I think we should choose either initialized or instantiated as the term we want to consistently use, rather than using both. I don't mind either term - whichever Ruby docs use would be best if anyone has a reference?

Reviewer 2: I agree with one or another.  If it is introductory, it may be OK to define via context one using the other.  Since we use initialize to instantiate an object, I believe the term is used to describe how we get an instance of a class.

Reviewer 2: This brings up the focus as well, Classes are initialized using `class`, while we get instances of that class by using `new`.  Perhaps the subject in that last sentence is the wrong noun.

Reviewer 3: We could use:
"Instances of a class are instantiated using the `new` method"
But I think I'd prefer:
"Classes are instantiated using the `new` method."

Author: @pvcarrera Yeah - do that second one.

Reviewer 2: Objects are instantiated by using the `new` method, classes are instantiated in other ways.

We create a class usually by doing

```ruby
class MyClass
end
```
This "instantiates a class" by creating a new object called `MyClass`.  We should be careful of mixing terms.  We instantiate objects of the class `MyClass` by doing `MyClass.new`

The instantiation of a class can be done by doing `MyClass = Class.new` which would be applying the idea directly, using the same terms as we would use when we generally talk about objects in Ruby.

Author: ```suggestion
Ruby is organised into classes. Classes are defined using the `class` keyword followed by the name of the class. Objects are created by instantiating classes using the `.new` method. For example:
```

Reviewer 2: "Objects are created by instantiating classes" is true... but it is only sometimes that the `.new` method is used.  This is immediately problematic as we start instantiating strings in examples, such as `fruit = 'banana'` where we have just instantiated a String object without using `.new`.  Or even the lie, when we use value objects, such as `the_answer = 42` and we get an `undefined method `new` for Integer:class`.

Instead:

    Objects are sometimes created by instnatiating classes using the `.new` method.

```ruby
# Define the class
class Calculator
  #...
end

# Create an instance of it and assign it to a variable
my_first_calc = Calculator.new
```

> A function within a class is referred to as a _method_. Each method can have zero or more parameters. Objects are returned from methods using the `return` keyword. If no return value is specified, the final object in the method is returned instead. Methods can also have named parameters which are defined and called using the `:` syntax.  Methods are invoked using `.` syntax.

Reviewer 2: ```suggestion
What may be called a _function_ in other languages are referred to as a _method_. Each method may have zero or more arguments. Objects may be returned from methods using the `return` keyword. If no return is specified, the result of the last evaluated statement of the method is returned instead. Methods can also have named parameters which are defined and called using the `:` syntax.  Methods are invoked using `.` syntax.
```

Reviewer 2: The "Methods can also have named parameters" section should be its own paragraph, where we can talk about artiy, and method signatures.  I.E., positional arguments, default arguments, keyword arguments, etc.

Reviewer: I disagree here. The exercise is not intended to cover everything there is to know about parameters and methods, but just to introduce some basics. To me, this excludes the things you're mentioning, which should IMHO be explained in a separate exercise.

Author: Yeah, I agree with Erik. We'll have another exercise that goes into the multitude of different ways to define methods and arguments. For this first exercise, I think showing the two most common approaches is enough I think, as the primary aim of the exercise is just to give students knowledge so that future exercises to be able to define methods using these ways.

Reviewer 2: I agree with Erik as well, in that that regard. Yes, its own paragraph, but perhaps not in this document.  The real problem I have is referring to the argument list as "parameters".

Author: @kotp OK - I'll reword to take that into account - thanks. Before I do though, can you explain why you think it should be `may` not `can`? The aim is to describe what is possible (e,g, it is possible to return from a method) rather than implying that someone has permission to do something, so I feel like "can" is grammatically correct here.

Reviewer 2: Because you may return objects using the `return` but it is not required.  Ruby `will` return an object, but you may do so either explicitly or implicitly.  (The word "can" implies that you are allowed to, while that isn't quite applicable here, you may do so, and loosely, yes, you "can" but you will do so, but you may do so in various ways.)

Does that make sense?

Reviewer 2: I thought about it a bit, and have written up this to hopefully clarify what my thoughts are about the words Methods and Functions.  While giving an example of an article that supports this, and noting that the article also talks about arguments and parameters in a way that I don't necessarily feel fits here as well.  (We can agree to use loose terms when needed, after all.)

---

What may be called a _function_ in other languages are referred to as a
_method_.<sup>[1]</sup>

Each method may have zero or more arguments. A Ruby method is allowed to
return one object.  We may return an object from a method by being
explicit by using the keyword `return`.  This is generally reserved for
when we need to return early from a method.  If no return is specified,
then the result of the last evaluated statement in the method will be
returned.  We call this an implicit return.

In Ruby, there really are no functions, if we consider that functions
are effected only by input, by the argument that is given.  However, we
do treat some methods similarly.  An example of this would be the
apparent use of `puts` or `print` in a void context.  Those methods will
return `nil` and appears to only be effected by the input to those
methods, effecting the output to a screen for example.

Methods are generally thought of to operate based on context, related to
the object on which they are called.

Note: While I agree with the "Methods Vs. Functions" portion of the
article, I don't agree with their definition of "Arguments Vs.
Parameters" in the strictest of senses, though I agree with the idea,
and understand when folks do reference them in the described way.
Sometimes we are a bit loose with our language when talking about
things, and I get that.  We tend to understand.  I try to be careful
while instructing though, since while instructing we tend to want to be
clear, while allowing for some "white lies" for a little bit of wiggle
room, to allow for understanding based on prior knowledge.

[1]: Methods Vs. Functions https://www.ombulabs.com/blog/ruby/learning/method-s-arguments-pt-1.html

---

I hope that gives some explanation why I reviewed this portion the way I did.

Author: Thanks @kotp. I appreciate the notes.

Re the "can" vs "may" I think in this situation we should use "can" as we are explaining what is possible, rather than what a user may choose to do. I understand that either are valid depending on how the sentence is aimed to be interpreted though.

How is this for a rewrite?

```suggestion
Units of functionality are encapsulated in methods - similar to _functions_ in other languages. A method can optionally be defined with positional arguments, or keyword arguments that are defined and called using the `:` syntax.  Methods either implicitly return the result of the last evaluated statement, or can explicitly return an object via the `return` keyword.  Methods are invoked using `.` syntax.
```

Reviewer 2: I like it... the one word "or" should be "and/or" for being able to use positional arguments and/or keyword arguments.  We can use one, or the other, or both.

Author: Added and/or. I'm going to stick with defined as we use `def` as the keywork and https://docs.ruby-lang.org/en/2.3.0/syntax/methods_rdoc.html exclusively uses the language of defining.

Thanks for all that @kotp.
